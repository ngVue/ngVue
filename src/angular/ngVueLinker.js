import angular from 'angular'
import Vue from 'vue'
import getVueComponent from '../components/getVueComponent'
import getPropExprs from '../components/props/getExpressions'
import watchPropExprs from '../components/props/watchExpressions'
import evalValues from '../components/props/evaluateValues'
import evalPropEvents from '../components/props/evaluateEvents'
import evaluateDirectives from '../directives/evaluateDirectives'
import extractSpecialAttributes from '../components/props/extractSpecialAttributes'
import watchSpecialAttributes from '../components/props/watchSpecialAttributes'

export function ngVueLinker (componentName, jqElement, elAttributes, scope, $injector) {
  if (!jqElement.parent().length) throw new Error("ngVue components must have a parent tag or they will not render")

  const $ngVue = $injector.has('$ngVue') ? $injector.get('$ngVue') : null
  const $compile = $injector.get('$compile')

  const dataExprsMap = getPropExprs(elAttributes)
  const Component = getVueComponent(componentName, $injector)
  const directives = evaluateDirectives(elAttributes, scope) || []
  const reactiveData = {
    _v: {
      props: evalValues(dataExprsMap.props || dataExprsMap.data, scope) || {},
      attrs: evalValues(dataExprsMap.htmlAttributes, scope) || {},
      special: extractSpecialAttributes(elAttributes)
    }
  }
  const on = evalPropEvents(dataExprsMap, scope) || {}

  const inQuirkMode = $ngVue ? $ngVue.inQuirkMode() : false
  const rootProps = $ngVue ? $ngVue.getRootProps() : {}

  const mounted = rootProps.mounted
  rootProps.mounted = function () {
    if (jqElement[0].innerHTML.trim()) {
      const html = $compile(jqElement[0].innerHTML)(scope)
      const slot = this.$refs.__slot__
      slot.parentNode.replaceChild(html[0], slot)
    }
    if (angular.isFunction(mounted)) {
      mounted.apply(this, arguments)
    }
  }

  const watchOptions = {
    depth: elAttributes.watchDepth,
    quirk: inQuirkMode
  }
  watchPropExprs(dataExprsMap, reactiveData, watchOptions, scope, 'props')
  watchPropExprs(dataExprsMap, reactiveData, watchOptions, scope, 'attrs')
  watchSpecialAttributes(reactiveData, jqElement, scope)

  let vueInstance = new Vue({
    name: 'NgVue',
    el: jqElement[0],
    data: reactiveData,
    render (h) {
      return (
        <Component
          {...{ directives }}
          {...{ props: reactiveData._v.props, on, attrs: reactiveData._v.attrs }}
          {...reactiveData._v.special}
        >
          {<span ref="__slot__" />}
        </Component>
      )
    },
    ...rootProps
  })

  scope.$on('$destroy', () => {
    vueInstance.$destroy()
    angular.element(vueInstance.$el).remove()
    vueInstance = null
  })
}
