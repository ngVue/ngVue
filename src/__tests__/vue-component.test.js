import angular from 'angular'
import Vue from 'vue'

import ngHtmlCompiler from './utils/ngHtmlCompiler'

import HelloComponent from './fixtures/HelloComponent'
import HelloWrappedComponent from './fixtures/HelloWrappedComponent'
import PersonsComponent from './fixtures/PersonsComponent'
import ButtonComponent from './fixtures/ButtonComponent'
import GreetingsComponent from './fixtures/GreetingsComponent'

describe('vue-component', () => {
  let $provide
  let $rootScope
  let compileHTML

  beforeEach(() => {
    angular.mock.module('ngVue')

    angular.mock.module((_$provide_) => {
      $provide = _$provide_
    })

    angular.mock.inject((_$rootScope_, _$compile_) => {
      $rootScope = _$rootScope_
      compileHTML = ngHtmlCompiler(_$rootScope_, _$compile_)
    })
  })

  describe('creation', () => {
    beforeEach(() => {
      $provide.value('HelloComponent', HelloComponent)
    })

    it('should render a vue component with name', () => {
      const elem = compileHTML('<vue-component name="HelloComponent" />')
      expect(elem[0].innerHTML.replace(/\s/g, '')).toBe('<span>Hello</span>')
    })

    it('should render a vue component with v-props object from scope', () => {
      const scope = $rootScope.$new()
      scope.person = { firstName: 'John', lastName: 'Doe' }
      const elem = compileHTML('<vue-component name="HelloComponent" v-props="person" />', scope)
      expect(elem[0].innerHTML).toBe('<span>Hello John Doe</span>')
    })

    it('should render a vue component with v-props-name properties from scope', () => {
      const scope = $rootScope.$new()
      scope.person = { firstName: 'John', lastName: 'Doe' }
      const elem = compileHTML(
        `<vue-component
          name="HelloComponent"
          v-props-first-name="person.firstName"
          v-props-last-name="person.lastName" />`,
        scope
      )
      expect(elem[0].innerHTML).toBe('<span>Hello John Doe</span>')
    })

    it('should render a vue component with original html attributes ', () => {
      const elem = compileHTML(
        `<vue-component
          name="HelloComponent"
          random="'hello'"
          tabindex="1"
          class="foo"
          style="font-size: 2em;"
          disabled
          data-qa="'John'" />`
      )
      expect(elem[0].innerHTML).toBe('<span random="hello" tabindex="1" disabled="disabled" data-qa="John" class="foo" style="font-size: 2em;">Hello  </span>')
    })

    it('should render a vue component with original html attributes on elements that bind $attrs ', () => {
      $provide.value('HelloWrappedComponent', HelloWrappedComponent)
      const elem = compileHTML(
        `<vue-component
          name="HelloWrappedComponent"
          random="'hello'"
          tabindex="1"
          class="foo"
          style="font-size: 2em;"
          disabled
          data-qa="'John'" />`
      )
      expect(elem[0].innerHTML).toBe('<div class="foo" style="font-size: 2em;"><span random="hello" tabindex="1" disabled="disabled" data-qa="John">Hello  </span></div>')
    })
  })

  describe('update', () => {
    beforeEach(() => {
      $provide.value('HelloComponent', HelloComponent)
      $provide.value('PersonsComponent', PersonsComponent)
    })

    it('should re-render the vue component when v-props value changes', (done) => {
      const scope = $rootScope.$new()
      scope.person = { firstName: 'John', lastName: 'Doe' }
      const elem = compileHTML('<vue-component name="HelloComponent" v-props="person" />', scope)

      scope.person.firstName = 'Jane'
      scope.person.lastName = 'Smith'
      Vue.nextTick(() => {
        expect(elem[0].innerHTML).toBe('<span>Hello Jane Smith</span>')
        done()
      })
    })

    it('should re-render the vue component when v-props reference changes', (done) => {
      const scope = $rootScope.$new()
      scope.person = { firstName: 'John', lastName: 'Doe' }
      const elem = compileHTML('<vue-component name="HelloComponent" v-props="person" />', scope)

      scope.person = { firstName: 'Jane', lastName: 'Smith' }
      scope.$digest()
      Vue.nextTick(() => {
        expect(elem[0].innerHTML).toBe('<span>Hello Jane Smith</span>')
        done()
      })
    })

    it('should re-render the vue component when v-props-name value change', (done) => {
      const scope = $rootScope.$new()
      scope.person = { firstName: 'John', lastName: 'Doe' }
      const elem = compileHTML(
        `<vue-component
          name="HelloComponent"
          v-props-first-name="person.firstName"
          v-props-last-name="person.lastName" />`,
        scope
      )

      scope.person.firstName = 'Jane'
      scope.person.lastName = 'Smith'
      scope.$digest()
      Vue.nextTick(() => {
        expect(elem[0].innerHTML).toBe('<span>Hello Jane Smith</span>')
        done()
      })
    })

    it('should re-render the vue component when v-props-name reference change', (done) => {
      const scope = $rootScope.$new()
      scope.person = { firstName: 'John', lastName: 'Doe' }
      const elem = compileHTML(
        `<vue-component
          name="HelloComponent"
          v-props-first-name="person.firstName"
          v-props-last-name="person.lastName" />`,
        scope
      )

      scope.person = { firstName: 'Jane', lastName: 'Smith' }
      scope.$digest()
      Vue.nextTick(() => {
        expect(elem[0].innerHTML).toBe('<span>Hello Jane Smith</span>')
        done()
      })
    })

    it('should re-render the vue component when v-props-name is an array and its items change', (done) => {
      const scope = $rootScope.$new()
      scope.persons = [
        { firstName: 'John', lastName: 'Doe' },
        { firstName: 'Jane', lastName: 'Doe' }
      ]
      const elem = compileHTML(
        `<vue-component
          name="PersonsComponent"
          v-props-persons="persons" />`,
        scope
      )

      // use Array.prototype.splice
      scope.persons.splice(0, 1, { firstName: 'John', lastName: 'Smith' })
      // use Vue.set
      Vue.set(scope.persons, 1, { firstName: 'Jane', lastName: 'Smith' })

      scope.$digest()
      Vue.nextTick(() => {
        expect(elem[0].innerHTML).toBe('<ul><li>John Smith</li><li>Jane Smith</li></ul>')
        done()
      })
    })

    it('should re-render a vue component with attribute values change', (done) => {
      const scope = $rootScope.$new()
      scope.isDisabled = false
      scope.tabindex = 0
      scope.randomAttr = "enabled"
      scope.class = 'foo'
      scope.size = '2em'
      const elem = compileHTML(
        `<vue-component
          name="HelloComponent"
          random="randomAttr"
          tabindex="tabindex"
          class="{{class}}"
          ng-style="{'font-size': size}"
          disabled="isDisabled" />`,
        scope
      )
      $rootScope.$digest()  // extra full digest needed for $animate to apply new class

      Vue.nextTick(() => {  // wait a tick for ng-* directives' settled values to propagate
        expect(elem[0].innerHTML).toBe('<span random="enabled" tabindex="0" class="foo" style="font-size: 2em;">Hello  </span>')

        scope.isDisabled = true
        scope.tabindex = 1
        scope.randomAttr = "disabled"
        scope.class = 'bar'
        scope.size = '3em'
        scope.$digest()
        $rootScope.$digest()  // extra full digest needed for $animate to apply new class

        Vue.nextTick(() => {
          expect(elem[0].innerHTML).toBe('<span random="disabled" tabindex="1" class="bar" style="font-size: 3em;" disabled="disabled">Hello  </span>')
          done()
        })
      })
    })
  })

  describe('remove', () => {
    beforeEach(() => {
      $provide.value('HelloComponent', HelloComponent)
    })

    it('should remove a vue component when ng-if directive flag toggles from true to false', () => {
      const scope = $rootScope.$new()
      scope.visible = true
      const elem = compileHTML(
        `<vue-component
          name="HelloComponent"
          ng-if="visible" />`,
        scope
      )
      expect(elem[0]).toMatchSnapshot()

      scope.visible = false
      scope.$digest()
      expect(elem[0]).toMatchSnapshot()
    })
  })

  describe('events', () => {
    it('should handle custom events from Vue', () => {
      $provide.value('ButtonComponent', ButtonComponent)

      const scope = $rootScope.$new()
      scope.handleHelloEvent = jest.fn()

      const elem = compileHTML(
        `<vue-component name="ButtonComponent" v-on-hello="handleHelloEvent" />`,
        scope
      )
      elem.find('button')[0].click()
      scope.$digest()
      expect(scope.handleHelloEvent).toHaveBeenCalledWith('Hello, World!')
    })
  })

  describe('slots', () => {
    beforeEach(() => {
      $provide.value('GreetingsComponent', GreetingsComponent)
    })

    it('should render a vue component with a button in the slot content', () => {
      const scope = $rootScope.$new()
      scope.onClick = jest.fn()

      const elem = compileHTML(`
        <vue-component name="GreetingsComponent">
          <button ng-click="onClick()">Click me!</button>
        </vue-component>`,
        scope
      )
      expect(elem[0]).toMatchSnapshot()

      elem.find('button')[0].click()
      scope.$digest()
      expect(scope.onClick).toHaveBeenCalled()
    })
  })
})
