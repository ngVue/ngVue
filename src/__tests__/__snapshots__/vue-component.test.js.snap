// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`vue-component creation should throw an error if there is no parent tag 1`] = `"ngVue components must have a parent tag or they will not render"`;

exports[`vue-component remove should remove a vue component when ng-if directive flag toggles from true to false 1`] = `
<div
  class="ng-scope"
>
  <!-- ngIf: visible -->
  <span>
    Hello  
  </span>
  <!-- end ngIf: visible -->
</div>
`;

exports[`vue-component remove should remove a vue component when ng-if directive flag toggles from true to false 2`] = `
<div
  class="ng-scope"
>
  <!-- ngIf: visible -->
</div>
`;

exports[`vue-component slots should render a vue component with a button in the slot content 1`] = `
<div
  class="ng-scope"
>
  <div>
    <span>
      Welcome, John Doe!
    </span>
    <button
      ng-click="onClick()"
      class="ng-scope"
    >
      Click me!
    </button>
  </div>
</div>
`;
